### PR TITLE
Signing: enable independent verification of primary signature and countersignature

### PR DESCRIPTION
**Pull request by dtivel**
Resolve https://github.com/NuGet/Home/issues/6993.

* Update `SignedPackageVerifierSettings` and `SignatureTrustAndValidityVerificationProvider` to allow verification of only the primary signature, only the repository countersignature, or both.
* Remove `RepositorySignatureVerifier`.
* Replace `SignatureVerificationStatus.Illegal` and `SignatureVerificationStatus.Untrusted` with `SignatureVerificationStatus.Disallowed`.  This simplifies outcomes while removing dependence of `treatIssuesAsErrors` on `allowIllegal`.
* Fix a couple of small bugs found while working on this.

CC @loic-sharma, @PatoBeltran

### CODE REVIEW

File: src/NuGet.Core/NuGet.Packaging/Signing/Signatures/Signature.cs [hunk 1]
```diff
@@ -188,8 +188,13 @@ protected Signature(SignerInfo signerInfo, SignatureType type)
 
                     CertificateChainUtility.SetCertBuildChainPolicy(chain.ChainPolicy, certificateExtraStore, timestamp.UpperLimit.LocalDateTime, CertificateType.Signature);
                     var chainBuildingSucceeded = CertificateChainUtility.BuildCertificateChain(chain, certificate, out var chainStatuses);
+                    var x509ChainString = CertificateUtility.X509ChainToString(chain, fingerprintAlgorithm);
+
+                    if (!string.IsNullOrWhiteSpace(x509ChainString))
```
Comment (dtivel) [inline src/NuGet.Core/NuGet.Packaging/Signing/Signatures/Signature.cs:193]: If a chain had only a single certificate, `CertificateUtility.X509ChainToString(...)` would return an empty string, which would get logged --- a verbose log with no message.

File: test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/AllowListVerificationProviderTests.cs [hunk 2]
```diff
@@ -1015,24 +1012,69 @@ public AllowListVerificationProviderTests(SigningTestFixture fixture)
                         signedPackageVerifier,
                         signedPackageVerifierSettings);
 
-                    // Act
-                    Action action = async () => await PackageExtractor.ExtractPackageAsync(
```
Comment (dtivel) [inline test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/SigningTests/AllowListVerificationProviderTests.cs:1019]: An exception was thrown with the require mode policy, but the test was not failing because the exception was not observed.  `action` was not awaited.

File: src/NuGet.Core/NuGet.Packaging/Signing/Signatures/SignatureVerifySettings.cs [hunk 3]
```diff
@@ -4,20 +4,24 @@
...
+        public bool AllowIllegal { get; }
 
         /// <summary>
         /// Specifies that a signing certificate's chain that chains to an untrusted root is allowed
         /// </summary>
-        public bool AllowUntrustedRoot { get; }
+        public bool AllowUntrusted { get; }
+
+        /// <summary>
+        /// Indicates if untrusted status should be reported.
```
Comment (loic-sharma) [inline src/NuGet.Core/NuGet.Packaging/Signing/Signatures/SignatureVerifySettings.cs]: >reported [](start = 52, length = 8)

Maybe we should mention that these are reported to the `issues` parameter of `Signature.Verify`?

**Review by loic-sharma — COMMENTED** [review]
Looks great! Thanks @dtivel

**Review by loic-sharma — APPROVED** [review]

File: src/NuGet.Core/NuGet.Packaging/Signing/Signatures/Signature.cs [hunk 4]
```diff
@@ -264,17 +272,35 @@ protected Signature(SignerInfo signerInfo, SignatureType type)
                     }
                     else if (!isSignatureTimeValid)
                     {
-                        if (settings.LogOnSignatureExpired)
-                        {
-                            issues.Add(SignatureLog.Issue(settings.TreatIssuesAsErrors, NuGetLogCode.NU3011, string.Format(CultureInfo.CurrentCulture, Strings.VerifyError_SignatureNotTimeValid, FriendlyName)));
-                        }
+                        issues.Add(
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Signatures/Signature.cs]: Why remove the check for `LogOnSignatureExpired`? The idea to have this was that when you want to check for the countersignature when the primary signature we wouldn't log any error for this because we will be checking expiry against the coutnersignature's timestamp. Should we be using `ReportUntrusted` here too?

File: src/NuGet.Core/NuGet.Packaging/Signing/Signatures/Signature.cs [hunk 5]
```diff
@@ -264,17 +272,35 @@ protected Signature(SignerInfo signerInfo, SignatureType type)
...
                         flags |= SignatureVerificationStatusFlags.CertificateExpired;
                         expirationTime = DateTime.SpecifyKind(certificate.NotAfter, DateTimeKind.Local);
                     }
                 }
             }
 
-            var status = VerificationUtility.GetSignatureVerificationStatus(flags);
+            SignatureVerificationStatus status;
+
+            if (flags != SignatureVerificationStatusFlags.NoErrors && (flags & ~SignatureVerificationStatusFlags.Untrusted) == 0)
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Signatures/Signature.cs]: Why is this needed? why only for untrusted? Can you just update `VerificationUtility.GetSignatureVerificationStatus(flags)` to return disallowed if it is `untrusted` and `AllowUntrusted` is `false` or if it is `illegal` and `AllowIllegal` is `false`?

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs [hunk 6]
```diff
@@ -13,115 +13,236 @@ namespace NuGet.Packaging.Signing
...
+                    shouldVerifyRepositoryCountersignature = isRepositoryCountersignatureVerificationRequested &&
+                        repositoryCountersignatureExists;
+                    break;
+
+                case SignatureVerificationBehavior.IfExistsAndIsNecessary:
+                    shouldVerifyRepositoryCountersignature = isRepositoryCountersignatureVerificationRequested &&
+                        repositoryCountersignatureExists &&
+                        (primarySummary == null ||
+                            (primarySummary != null &&
+                            (PrimarySignatureHasUntrustedRoot(primarySummary) || PrimarySignatureNeedsTimestamp(primarySummary))));
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs]: Consider changing `PrimarySignatureNeedsTimestamp` to something like `PrimarySignatureExpired`. It can be the case that this returns true even if the primary signature already has a timestamp. Technically "needsTimestamp" would not be incorrect in that case, but I find it a bit confusing to state that it needs a timestamp.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs [hunk 7]
```diff
@@ -13,115 +13,236 @@ namespace NuGet.Packaging.Signing
...
+                VerificationUtility.IsVerificationTarget(signature.Type, settings.VerificationTarget))
             {
+                primarySummary = VerifyValidityAndTrust(signature, settings, verifySettings, certificateExtraStore, issues);
+
                 status = primarySummary.Status;
+            }
+
+            bool shouldVerifyRepositoryCountersignature;
+
+            switch (settings.RepositoryCountersignatureVerificationBehavior)
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs]: should we fail if `isRepositoryCountersignatureVerificationRequested` is `false` in any case exept `never`?

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs [hunk 8]
```diff
@@ -13,115 +13,236 @@ namespace NuGet.Packaging.Signing
...
+                            }
 
-                        if (!Rfc3161TimestampVerificationUtility.ValidateSignerCertificateAgainstTimestamp(signature.SignerInfo.Certificate, counterSummary.Timestamp))
+                            status = SignatureVerificationStatus.Valid;
+                        }
+                        else
                         {
-                            issues.Add(SignatureLog.Issue(!settings.AllowIllegal, NuGetLogCode.NU3011, string.Format(CultureInfo.CurrentCulture, Strings.VerifyError_SignatureNotTimeValid, signature.FriendlyName)));
-                            status = SignatureVerificationStatus.Illegal;
+                            issues.Add(
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs]: Given that, if the primary signature is expired it will log an issue, wouldn't this issue be logged twice?

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignedPackageVerifierSettings.cs [hunk 9]
```diff
@@ -105,22 +128,82 @@ public sealed class SignedPackageVerifierSettings
...
+            {
+                throw new ArgumentException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.UnrecognizedEnumValue,
+                        repositoryCountersignatureVerificationBehavior),
+                    nameof(repositoryCountersignatureVerificationBehavior));
+            }
+
+            if ((signaturePlacement.HasFlag(SignaturePlacement.Countersignature) && !verificationTarget.HasFlag(VerificationTarget.Repository)) ||
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignedPackageVerifierSettings.cs:170]: 🥇

**Review by PatoBeltran — COMMENTED** [review]
Just some comments, in general it is looking good! 🌵

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs (hunk 7 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs]: `SignedPackageVerifierSettings` does not allow that combination of values.  I've added a debug assertion.

File: src/NuGet.Core/NuGet.Packaging/Signing/Utility/VerificationUtility.cs [hunk 10]
```diff
@@ -14,27 +14,43 @@ public static class VerificationUtility
...
 
-            if ((flags & SignatureVerificationStatusFlags.Illegal) != 0)
+            if ((flags & SignatureVerificationStatusFlags.Suspect) != 0)
             {
-                return SignatureVerificationStatus.Illegal;
+                return SignatureVerificationStatus.Suspect;
             }
 
-            if ((flags & SignatureVerificationStatusFlags.Untrusted) != 0)
+            if ((flags & ~(SignatureVerificationStatusFlags.Illegal |
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging/Signing/Utility/VerificationUtility.cs]: it would be helpful to have a comment explaining this condition

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/PackageSignatureVerifier.cs [hunk 11]
```diff
@@ -102,12 +108,9 @@ public PackageSignatureVerifier(IEnumerable<ISignatureVerificationProvider> veri
         private static bool IsValid(IEnumerable<PackageVerificationResult> verificationResults, SignedPackageVerifierSettings settings)
         {
             var hasItems = verificationResults.Any();
-            var valid = verificationResults.All(e =>
-                e.Trust == SignatureVerificationStatus.Valid ||
-                (settings.AllowIllegal && SignatureVerificationStatus.Illegal == e.Trust) ||
-                (settings.AllowUntrusted && SignatureVerificationStatus.Untrusted == e.Trust));
+            var allValid = verificationResults.All(e => e.Trust == SignatureVerificationStatus.Valid);
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/PackageSignatureVerifier.cs]: shouldn't this be conditioned on hasItems?

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs [hunk 12]
```diff
@@ -13,115 +15,234 @@ namespace NuGet.Packaging.Signing
...
+
+            switch (settings.RepositoryCountersignatureVerificationBehavior)
+            {
+                case SignatureVerificationBehavior.IfExists:
+                    shouldVerifyRepositoryCountersignature = isRepositoryCountersignatureVerificationRequested &&
+                        repositoryCountersignatureExists;
+                    break;
+
+                case SignatureVerificationBehavior.IfExistsAndIsNecessary:
+                    shouldVerifyRepositoryCountersignature = isRepositoryCountersignatureVerificationRequested &&
```
Comment (rohit21agrawal) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs]: a comment to explain the condition would be appreciated here too

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs [hunk 13]
```diff
@@ -13,115 +15,234 @@ namespace NuGet.Packaging.Signing
...
-                            issues.Add(SignatureLog.Issue(!settings.AllowIllegal, NuGetLogCode.NU3011, string.Format(CultureInfo.CurrentCulture, Strings.VerifyError_SignatureNotTimeValid, signature.FriendlyName)));
-                            status = SignatureVerificationStatus.Illegal;
+                            // Exclude the issue of the primary signature being expired since the repository countersignature fulfills the role of a trusted timestamp.
+                            issues = issues.Where(log => log.Code != NuGetLogCode.NU3037);
+
+                            status = SignatureVerificationStatus.Valid;
                         }
                     }
-                    else if (primarySummary.Flags.HasFlag(SignatureVerificationStatusFlags.CertificateExpired))
+                    else if (countersignatureSummary.Status == SignatureVerificationStatus.Valid &&
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs]: Can there be a case where the primary signature is expired and also has untrusted root? Should this be an `if` instead of an `else if`?

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureVerificationSummary.cs [hunk 14]
```diff
@@ -30,38 +31,44 @@ public sealed class SignatureVerificationSummary
         /// <summary>
         /// Expiration Date and Time for signature
         /// </summary>
-        /// <remarks>This field will only be set if the flag CertificateExpired is present <see cref="Flags" /></remarks>
+        /// <remarks>This field will only be set if the certificate is expired.</remarks>
         public DateTimeOffset? ExpirationTime { get; }
 
+        public IEnumerable<SignatureLog> Issues { get; set; }
```
Comment (PatoBeltran) [inline src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureVerificationSummary.cs:37]: What made you decide on moving the `issues` to this class?

**Review by PatoBeltran — APPROVED** [review]
Only two small questions. It looks awesome! 👍

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureVerificationSummary.cs (hunk 14 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureVerificationSummary.cs:37]: Two reasons:

1.  Issues found are arguably part of the verification summary.
2.  There are two cases where issues reported for the primary signature can be removed if a repository countersignature is used as a fallback.  In those cases we need to be certain that when remove an issue it's being removed only from the set of issues for the primary signature.  With the previous approach of accumulating issues in a single `List<SignatureLog>`, after we evaluate the primary signature and the repository countersignature, the list contains issues for both signatures.  Removing an issue and then concatenating issues seems safer than concatenating issues and then removing an issue.

File: src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs (hunk 13 above)
Comment (dtivel) [reply src/NuGet.Core/NuGet.Packaging/Signing/Verification/SignatureTrustAndValidityVerificationProvider.cs]: Good catch!
